webhook/resourcesemantics/validation/validation.go

```diff
@@ -238,21 +245,25 @@ func (ac *reconciler) validate(ctx context.Context, req *admissionv1beta1.Admiss
 		return errMissingNewObject
 	}
 
-	if err := validate(ctx, newObj); err != nil {
+	if err := newObj.Validate(ctx); err != nil {
 		logger.Errorw("Failed the resource specific validation", zap.Error(err))
 		// Return the error message as-is to give the validation callback
 		// discretion over (our portion of) the message that the user sees.
 		return err
 	}
 
-	return nil
-}
+	// Generically callback if any are provided for the resource.
+	if callback, ok := ac.callbacks[gvk]; ok {
+		uns, err := runtime.DefaultUnstructuredConverter.ToUnstructured(newObj)
+		if err != nil {
+			return fmt.Errorf("cannot convert new object to unstructured: %v", err)
+		}
+		unstruct := &unstructured.Unstructured{}
+		unstruct.SetUnstructuredContent(uns)
```

Comment:
Given that we're not dealing with mutations, I bet it is more efficient to simply decode the raw object from `newBytes` into an unstructured, but otherwise this seems reasonable to me.

Reply:
https://github.com/knative/pkg/pull/1177

Sure I've written a benchmark to test those two options. You're right - straight decoding is faster.

Output of 10,000 trials:
Converting took: 315.162774ms
Decoding took: 263.158939ms